pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,351 @@
+func (c *MultitenantCompactor) UploadBlockFile(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	blockID := vars["block"]
+	if blockID == "" {
+		http.Error(w, "missing block ID", http.StatusBadRequest)
+		return
+	}
+	_, err := ulid.Parse(blockID)
+	if err != nil {
+		http.Error(w, "invalid block ID", http.StatusBadRequest)
+		return
+	}
+	pth := r.URL.Query().Get("path")
+	if pth == "" {
+		http.Error(w, "missing or invalid file path", http.StatusBadRequest)
+		return
+	}
+
+	tenantID, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+	if err != nil {
+		http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+		return
+	}
+
+	logger := util_log.WithContext(ctx, c.logger)
+	logger = log.With(logger, "block", blockID)
+
+	if path.Base(pth) == block.MetaFilename {
+		http.Error(w, fmt.Sprintf("%s is not allowed", block.MetaFilename), http.StatusBadRequest)
+		return
+	}
+
+	if !rePath.MatchString(pth) {
+		http.Error(w, fmt.Sprintf("invalid path: %q", pth), http.StatusBadRequest)
+		return
+	}
+
+	if r.ContentLength == 0 {
+		http.Error(w, "file cannot be empty", http.StatusBadRequest)
+		return
+	}
+
+	userBkt := bucket.NewUserBucketClient(tenantID, c.bucketClient, c.cfgProvider)
+
+	metaPath := path.Join(blockID, tmpMetaFilename)
+	exists, err := userBkt.Exists(ctx, metaPath)
+	if err != nil {
+		level.Error(logger).Log("msg", fmt.Sprintf("failed to check existence of %s in object storage", tmpMetaFilename),
+			"err", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+	if !exists {
+		http.Error(w, fmt.Sprintf("upload of block %s not started yet", blockID), http.StatusBadRequest)
+		return
+	}
+
+	// TODO: Verify that upload path and length correspond to file index
+
+	dst := path.Join(blockID, pth)
+
+	level.Debug(logger).Log("msg", "uploading block file to bucket", "destination", dst,
+		"size", r.ContentLength)
+	reader := bodyReader{
+		r: r,
+	}
+	if err := userBkt.Upload(ctx, dst, reader); err != nil {
+		level.Error(logger).Log("msg", "failed uploading block file to bucket",
+			"destination", dst, "err", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	level.Debug(logger).Log("msg", "finished uploading block file to bucket",
+		"path", pth)
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func decodeMeta(r io.Reader, name string) (metadata.Meta, error) {
+	dec := json.NewDecoder(r)
+	var meta metadata.Meta
+	if err := dec.Decode(&meta); err != nil {
+		return meta, errors.Wrap(err, fmt.Sprintf("failed decoding %s", name))
+	}
+
+	return meta, nil
+}
+
+func (c *MultitenantCompactor) completeBlockUpload(ctx context.Context, r *http.Request,
+	logger log.Logger, userBkt objstore.Bucket, tenantID string, blockID ulid.ULID) error {
+	level.Debug(logger).Log("msg", "received request to complete block upload", "content_length", r.ContentLength)
+
+	rdr, err := userBkt.Get(ctx, path.Join(blockID.String(), tmpMetaFilename))
+	if err != nil {
```

Comment:
@pstibrany good observation, fixing. I'm also changing the handler for uploading block files, to return 404 to the user if the file is missing (instead of 400) to be consistent.

Reply:
Thanks!